x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/CachingServiceAccountsTokenStore.java

```diff
@@ -112,7 +121,17 @@ public final void invalidate(Collection<String> qualifiedTokenNames) {
         if (cache != null) {
             logger.trace("invalidating cache for service token [{}]",
                 Strings.collectionToCommaDelimitedString(qualifiedTokenNames));
-            qualifiedTokenNames.forEach(cache::invalidate);
+            if (qualifiedTokenNames.size() == 1) {
+                final String qualifiedTokenName = qualifiedTokenNames.iterator().next();
+                if (qualifiedTokenName.endsWith("/")) {
+                    // Wildcard case of invalidating all tokens for a service account, e.g. "elastic/fleet-server/"
+                    cacheIteratorHelper.removeKeysIf(key -> key.startsWith(qualifiedTokenName));
+                } else {
+                    cache.invalidate(qualifiedTokenName);
+                }
+            } else {
+                qualifiedTokenNames.forEach(cache::invalidate);
+            }
```

Comment:
Why don't we implement the special trailing `/` behaviour when there are multiple token names?
I understand that the callers never use that, but it seems strange to bake it into this API.

My rule of thumb here is, if I had to describe the behaviour in javadoc, would it make sense?
I think saying that the special case only applies if the argument contains a single entry, implies that either:
1. This should be 2 different methods, _or_
2. We should just do that check for every element in the collection.

Reply:
This is done to match the existing behaviour of other cache clearing APIs. It may not be obvious due to the extra `namespace` and `service` fields. 

Using the clear roles cache API as an example, the request `POST _security/role/*,foo/_clear_cache` does not interpret the `*` as a wildcard but instead of a literal name. Similarly `POST _security/role/,foo/_clear_cache` treats the empty string literally as well. You only get the wildcard behaviour if the rquest is `POST _security/role/*/_clear_cache`.

`TransportClearRolesCacheAction` basically has the same behaviour with its [nodeOperation](https://github.com/elastic/elasticsearch/blob/7.12/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/role/TransportClearRolesCacheAction.java#L55-L64) method. The only difference here is that we need to check something like `elastic/fleet-server/` instead of an empty string because of the namespace and service prefix. If you see past them, it is essentially just an empty token name string.

Reply:
If you are unhappy with this approach, I can change it to always clear all tokens for a service account (the alternative you suggested above). The change will naturally get rid of this issue.

Reply:
Actually, it will not completely get rid of the issue because clearing a single key is still needed by deleting API-created keys. So the argument of `Collection<String> qualifiedTokenNames` will be either a singleton list of `elastic/fleet-server/` for clearing all tokens of this service account or a singleton list of `elastic/fleet-server/token_name` for clearing a single token.